README.md
# Merkle Tree Algorithm Implementation using PHP

[![Latest Version on Packagist](https://img.shields.io/packagist/v/mohamedahmed01/simple-merkele.svg?style=flat-square)](https://packagist.org/packages/mohamedahmed01/simple-merkele#v1.0.0beta1)
[![Build Status](https://travis-ci.com/mohamedahmed01/simple-merkele.svg?branch=main)](https://travis-ci.com/mohamedahmed01/simple-merkele)
[![Quality Score](https://img.shields.io/scrutinizer/g/mohamedahmed01/simple-merkele.svg?style=flat-square)](https://scrutinizer-ci.com/g/mohamedahmed01/simple-merkele)
[![Code Coverage](https://scrutinizer-ci.com/g/mohamedahmed01/simple-merkele/badges/coverage.png?b=main)](https://scrutinizer-ci.com/g/mohamedahmed01/simple-merkele/?branch=main)

This is a very simple implemenation of merkle tree in which you can use multiple hashes
to create a merkele tree and compress it to a single hash .

## Installation

You can install the package via composer:

```bash
composer require mohamedahmed01/simple-merkele
```

## Usage

``` php
    //include SimpleMereke
    //prepare your array of hashes
    $hashsArray=[
            hash('sha256', 'hello'),
            hash('sha256', 'goodbye'),
            hash('sha256', 'topofthemorning'),
            hash('sha256', 'sionara'),
            hash('sha256', 'areviditche'),
            hash('sha256', 'arregato'),
        ];
    //create new instance of simpleMerkele
    $merkele = new SimpleMerkele();
    //add your hashes to the inventory
    foreach ($hashsArray as $hash) {
        $merkele->addHash($hash);
    }
    //create the top hash
    $merkele->calculateTree(); //9b8dd5dd1f56d5fa17a67c10b8891c57e51f5fd36fe3a2d7e290d605840332d8
    $merkele->resetTree();
        
```

### Methods :

| Method | Description |
| --- | --- |
| SimpleMerkele($mode,$algo) | Creates new instance with the specified configuration |
| addHash($hash) | add hash to the internal inventory not that once added cannot be removed |
| calculateTree() | create the tree and the top node i.e hash |
| resetTree() | reset the internal memories to be able to add new hashes and create new tree |

### Configuration :

| Flag/Param | Description |
| --- | --- |
| SimpleMerkele::ALLOW_ODD_ARRAYS | Allow odd length trees but adjusting the pairing internally |
| SimpleMerkele::DIS_ALLOW_ODD_ARRAYS | throws an exceptions on odd length tree |
| $algo | accept standard algo type i.e "md5","sha256" refer to HashValidation class for full list |

### Testing

``` bash
composer test
```

### Security

If you discover any security related issues, please email mohamedabdelmenem01@gmail.com instead of using the issue tracker.


## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.



CHANGELOG.md
# Changelog

All notable changes to `simple-merkele` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


tests/ExampleTest.php
<?php

namespace Chromax\SimpleMerkele\Tests;

use PHPUnit\Framework\TestCase;

class ExampleTest extends TestCase
{
    /** @test */
    public function true_is_true()
    {
        $this->assertTrue(true);
    }
}


tests/SimpleMerkleTest.php
<?php

namespace Mohamedahmed01\SimpleMerkele\Tests;

use Mohamedahmed01\SimpleMerkele\Exceptions\HashInventoryEmptyException;
use PHPUnit\Framework\TestCase;
use Mohamedahmed01\SimpleMerkele\SimpleMerkele;
use Mohamedahmed01\SimpleMerkele\Exceptions\InvalidHashException;
use Mohamedahmed01\SimpleMerkele\Exceptions\InvalidHashTypeException;
use Mohamedahmed01\SimpleMerkele\Exceptions\InvalidArrayModeException;
use Mohamedahmed01\SimpleMerkele\Exceptions\OddHashInventoryException;

class SimpleMerkleTest extends TestCase
{
    /** @test */
    public function createSimpleMerkele()
    {
        $this->assertInstanceOf(SimpleMerkele::class, new SimpleMerkele());
    }

    /** @test */
    public function simpleMerkleThrowsExceptionOnInvalidArrayMode()
    {
        $this->expectException(InvalidArrayModeException::class);
        $this->assertInstanceOf(SimpleMerkele::class, new SimpleMerkele(5));
    }

    /** @test */
    public function simpleMerkleThrowsExceptionOnInvalidHashType()
    {
        $this->expectException(InvalidHashTypeException::class);
        $this->assertInstanceOf(SimpleMerkele::class, new SimpleMerkele(2, 'banana'));
    }

    /** @test */
    public function simpleMerkeleHasAddHash()
    {
        $this->assertTrue(
            method_exists(new SimpleMerkele(), 'addHash'),
            'Class does not have method addHash'
        );
    }

    /** @test */
    public function addHashThrowsExceptionOnInvalidHashLength()
    {
        $this->expectException(InvalidHashException::class);
        (new SimpleMerkele())->addHash('AAAA');
    }

    /** @test */
    public function addHashThrowsExceptionOnInvalidHashCharachter()
    {
        $this->expectException(InvalidHashException::class);
        (new SimpleMerkele())->addHash('وادع الصخور لترحم العظم المُهشم');
    }

    /** @test */
    public function simpleMerkeleHasCalulateTree()
    {
        $this->assertTrue(
            method_exists(new SimpleMerkele(), 'calculateTree'),
            'Class does not have method calculateTree'
        );
    }

    /** @test */
    public function calulateTreeThrowsExceptionOnEmptyInventory()
    {
        $this->expectException(HashInventoryEmptyException::class);
        (new SimpleMerkele())->calculateTree();
    }

    /** @test */
    public function calulateTreeThrowsExceptionOnInvalidInventoryUsingDisallowODDInventory()
    {
        $this->expectException(OddHashInventoryException::class);
        $merkele = new SimpleMerkele(SimpleMerkele::DIS_ALLOW_ODD_ARRAYS);
        $merkele->addHash(hash('sha256', 'hello'));
        $merkele->calculateTree();
    }

    /** @test */
    public function calulateTreeReturnsExpectedHash()
    {
        $hashsArray=[
                hash('sha256', 'hello'),
                hash('sha256', 'goodbye'),
                hash('sha256', 'topofthemorning'),
                hash('sha256', 'sionara'),
                hash('sha256', 'areviditche'),
                hash('sha256', 'arregato'),
            ];
        $merkele = new SimpleMerkele();
        foreach ($hashsArray as $hash) {
            $merkele->addHash($hash);
        }
        $this->assertEquals(
            '9b8dd5dd1f56d5fa17a67c10b8891c57e51f5fd36fe3a2d7e290d605840332d8',
            $merkele->calculateTree()
        );
    }

        /** @test */
        public function calulateTreeReturnsExpectedHashOnUsingOddArrays()
        {
            $hashsArray=[
                    hash('sha256', 'hello'),
                    hash('sha256', 'goodbye'),
                    hash('sha256', 'topofthemorning'),
                    hash('sha256', 'sionara'),
                    hash('sha256', 'areviditche'),
                ];
            $merkele = new SimpleMerkele(SimpleMerkele::ALLOW_ODD_ARRAYS);
            foreach ($hashsArray as $hash) {
                $merkele->addHash($hash);
            }
            $this->assertEquals(
                '77894a7ac0f394835a304b53023bd25bf6c645eab0b1ef81e6636cab971f47e4',
                $merkele->calculateTree()
            );
        }
    /** @test */
    public function simpleMerkeleHasResetTree()
    {
        $this->assertTrue(
            method_exists(new SimpleMerkele(), 'resetTree'),
            'Class does not have method resetTree'
        );
    }
    /** @test */
    public function resetTreeEmptiesTheInventory()
    {
        $this->expectException(HashInventoryEmptyException::class);
        $hashsArray=[
                hash('sha256', 'hello'),
                hash('sha256', 'goodbye'),
                hash('sha256', 'topofthemorning'),
                hash('sha256', 'sionara'),
                hash('sha256', 'areviditche'),
                hash('sha256', 'arregato'),
            ];
        $merkele = new SimpleMerkele();
        foreach ($hashsArray as $hash) {
            $merkele->addHash($hash);
        }

        $merkele->calculateTree();
        $merkele->resetTree();
        $merkele->calculateTree();
    }
}


tests/HashValidationTest.php
<?php

namespace Mohamedahmed01\SimpleMerkele\Tests;

use PHPUnit\Framework\TestCase;
use Mohamedahmed01\SimpleMerkele\Types\HashValidation;
use Mohamedahmed01\SimpleMerkele\Exceptions\InvalidHashException;
use Mohamedahmed01\SimpleMerkele\Exceptions\InvalidHashTypeException;

class HashValidationTest extends TestCase
{

    /** @test */
    public function hashValidationHasGetHashLength()
    {
        $this->assertTrue(
            method_exists(HashValidation::class, 'getHashLength'),
            'Class does not have method getHashLength'
        );
    }

    /** @test
    * @SuppressWarnings(PHPMD.StaticAccess)
    */
    public function getHashLengthReturnsTheCorrectHashLength()
    {
        $this->assertEquals(HashValidation::getHashLength('md4'), 32);
        $this->assertEquals(HashValidation::getHashLength('md5'), 32);
        $this->assertEquals(HashValidation::getHashLength('sha1'), 40);
        $this->assertEquals(HashValidation::getHashLength('sha224'), 56);
        $this->assertEquals(HashValidation::getHashLength('sha256'), 64);
        $this->assertEquals(HashValidation::getHashLength('sha384'), 96);
        $this->assertEquals(HashValidation::getHashLength('sha512/224'), 56);
        $this->assertEquals(HashValidation::getHashLength('sha512/256'), 64);
        $this->assertEquals(HashValidation::getHashLength('sha512'), 128);
        $this->assertEquals(HashValidation::getHashLength('sha3-224'), 56);
        $this->assertEquals(HashValidation::getHashLength('sha3-256'), 64);
        $this->assertEquals(HashValidation::getHashLength('sha3-384'), 96);
        $this->assertEquals(HashValidation::getHashLength('sha3-512'), 128);
        $this->assertEquals(HashValidation::getHashLength('ripemd128'), 32);
        $this->assertEquals(HashValidation::getHashLength('ripemd160'), 40);
        $this->assertEquals(HashValidation::getHashLength('ripemd256'), 64);
        $this->assertEquals(HashValidation::getHashLength('ripemd320'), 80);
        $this->assertEquals(HashValidation::getHashLength('whirlpool'), 128);
        $this->assertEquals(HashValidation::getHashLength('tiger128,3'), 32);
        $this->assertEquals(HashValidation::getHashLength('tiger160,3'), 40);
        $this->assertEquals(HashValidation::getHashLength('tiger192,3'), 48);
        $this->assertEquals(HashValidation::getHashLength('tiger128,4'), 32);
        $this->assertEquals(HashValidation::getHashLength('tiger160,4'), 40);
        $this->assertEquals(HashValidation::getHashLength('tiger192,4'), 48);
        $this->assertEquals(HashValidation::getHashLength('snefru'), 64);
        $this->assertEquals(HashValidation::getHashLength('snefru256'), 64);
        $this->assertEquals(HashValidation::getHashLength('gost'), 64);
        $this->assertEquals(HashValidation::getHashLength('gost-crypto'), 64);
        $this->assertEquals(HashValidation::getHashLength('adler32'), 8);
        $this->assertEquals(HashValidation::getHashLength('crc32'), 8);
        $this->assertEquals(HashValidation::getHashLength('crc32b'), 8);
        $this->assertEquals(HashValidation::getHashLength('crc32c'), 8);
        $this->assertEquals(HashValidation::getHashLength('fnv132'), 8);
        $this->assertEquals(HashValidation::getHashLength('fnv1a32'), 8);
        $this->assertEquals(HashValidation::getHashLength('fnv164'), 16);
        $this->assertEquals(HashValidation::getHashLength('fnv1a64'), 16);
        $this->assertEquals(HashValidation::getHashLength('joaat'), 8);
        $this->assertEquals(HashValidation::getHashLength('haval128,3'), 32);
        $this->assertEquals(HashValidation::getHashLength('haval160,3'), 40);
        $this->assertEquals(HashValidation::getHashLength('haval192,3'), 48);
        $this->assertEquals(HashValidation::getHashLength('haval224,3'), 56);
        $this->assertEquals(HashValidation::getHashLength('haval256,3'), 64);
        $this->assertEquals(HashValidation::getHashLength('haval128,4'), 32);
        $this->assertEquals(HashValidation::getHashLength('haval160,4'), 40);
        $this->assertEquals(HashValidation::getHashLength('haval192,4'), 48);
        $this->assertEquals(HashValidation::getHashLength('haval224,4'), 56);
        $this->assertEquals(HashValidation::getHashLength('haval256,4'), 64);
        $this->assertEquals(HashValidation::getHashLength('haval128,5'), 32);
        $this->assertEquals(HashValidation::getHashLength('haval160,5'), 40);
        $this->assertEquals(HashValidation::getHashLength('haval192,5'), 48);
        $this->assertEquals(HashValidation::getHashLength('haval224,5'), 56);
        $this->assertEquals(HashValidation::getHashLength('haval256,5'), 64);
    }


    /** @test
    * @SuppressWarnings(PHPMD.StaticAccess)
    */
    public function getHashLengthReturnsExceptionForInvalidTypes()
    {
        $this->expectException(InvalidHashTypeException::class);
        $this->assertEquals(HashValidation::getHashLength('banana'), 32);
    }

    /** @test */
    public function hashValidationHascheckHashLength()
    {
        $this->assertTrue(
            method_exists(HashValidation::class, 'checkHashLength'),
            'Class does not have method checkHashLength'
        );
    }

    /** @test */
    public function hashValidationHasValidateHashString()
    {
        $this->assertTrue(
            method_exists(HashValidation::class, 'validateHashString'),
            'Class does not have method validateHashString'
        );
    }

    /** @test
    * @SuppressWarnings(PHPMD.StaticAccess)
    */
    public function hashValidationDontRaiseExceptionOnValidString()
    {
        $this->assertNull(
            HashValidation::validateHashString('146a7492719b3564094efe7abbd40a7416fd900179d02773')
        );
    }

    /** @test
    * @SuppressWarnings(PHPMD.StaticAccess)
    */
    public function hashValidationDontRaiseExceptionOnInValidString()
    {
        $this->expectException(InvalidHashException::class);
        $this->assertNull(
            HashValidation::validateHashString('الكلب قفز فوق الثعلب')
        );
    }

    /** @test */
    public function hashValidationHasValidHashingAlgo()
    {
        $this->assertTrue(
            method_exists(HashValidation::class, 'validHashingAlgo'),
            'Class does not have method getHashLength'
        );
    }

    /** @test
    * @SuppressWarnings(PHPMD.StaticAccess)
    */
    public function validHashingAlgothrowsExceptionForInvalidTypes()
    {
        $this->expectException(InvalidHashTypeException::class);
        $this->assertNull(HashValidation::validHashingAlgo('banana'));
    }
}


src/SimpleMerkele.php
<?php

namespace Mohamedahmed01\SimpleMerkele;

use Mohamedahmed01\SimpleMerkele\Types\HashValidation;
use Mohamedahmed01\SimpleMerkele\Exceptions\InvalidArrayModeException;
use Mohamedahmed01\SimpleMerkele\Exceptions\OddHashInventoryException;
use Mohamedahmed01\SimpleMerkele\Exceptions\HashInventoryEmptyException;

final class SimpleMerkele
{
    private $inventory = [];
    private $topHash = '';
    private $array_mode = 2;
    private $hashing_algo = 'sha256';
    public const ALLOW_ODD_ARRAYS = 1;
    public const DIS_ALLOW_ODD_ARRAYS = 2;

    public function __construct($array_mode=2, $hashing_algo='sha256')
    {
        $this->validateArrayMode($array_mode);
        $this->array_mode = $array_mode;
        HashValidation::validHashingAlgo($hashing_algo);
        $this->hashing_algo=$hashing_algo;
    }
    
    public function addHash(String $hash)
    {
        $this->validateHash($hash, $this->hashing_algo);
        $this->inventory[]=$hash;
    }

    public function calculateTree():string
    {
        $this->validateInventory();
        $this->reduceTree($this->inventory, []);
        return $this->topHash;
    }

    public function resetTree()
    {
        $this->inventory=[];
        $this->topHash = '';
    }

    private function reduceTree($treeInventory, $nodeHashesMemo=[])
    {
        $nodeHashes=$nodeHashesMemo;
        if (count($treeInventory)<=0) {
            $treeInventory=$nodeHashes;
            $nodeHashes=[];
            if (count($treeInventory) <= 1 &&empty($nodeHashes)) {
                $this->topHash=array_pop($treeInventory);
                return;
            }
            if (!(count($treeInventory)%2===0)) {
                $treeInventory[]=$treeInventory[count($treeInventory)-1];
            }
        }
        $nodeHashes[] = hash($this->hashing_algo, array_pop($treeInventory).array_pop($treeInventory));
        $this->reduceTree($treeInventory, $nodeHashes);
    }

    /**
    * @SuppressWarnings(PHPMD.StaticAccess)
    */
    private function validateHash(String $hash, String $type)
    {
        HashValidation::validateHashString($hash);
        HashValidation::checkHashLength($hash, $type);
    }

    private function validateInventory()
    {
        if (empty($this->inventory)) {
            throw new HashInventoryEmptyException();
        }
        $isOddArray = count($this->inventory)%2 !== 0;
        if ($this->array_mode === self::ALLOW_ODD_ARRAYS && $isOddArray) {
            $this->inventory[]=$this->inventory[count($this->inventory)-1];
        } elseif ($this->array_mode === self::DIS_ALLOW_ODD_ARRAYS && $isOddArray) {
            throw new OddHashInventoryException();
        }
    }

    private function validateArrayMode($array_mode)
    {
        if (!in_array(
            $array_mode,
            [self::ALLOW_ODD_ARRAYS,self::DIS_ALLOW_ODD_ARRAYS],
            true
        )) {
            throw new InvalidArrayModeException();
        }
    }
}


src/Types/HashValidation.php
<?php
namespace Mohamedahmed01\SimpleMerkele\Types;

use Mohamedahmed01\SimpleMerkele\Exceptions\InvalidHashException;
use Mohamedahmed01\SimpleMerkele\Exceptions\InvalidHashTypeException;

final class HashValidation
{
    private function __construct()
    {
    }
    private const HASH_TYPES =[
        'md4'=> 32,
        'md5'=> 32,
        'sha1'=> 40,
        'sha224'=> 56,
        'sha256'=> 64,
        'sha384'=> 96,
        'sha512/224'=> 56,
        'sha512/256'=> 64,
        'sha512'=> 128,
        'sha3-224'=> 56,
        'sha3-256'=> 64,
        'sha3-384'=> 96,
        'sha3-512'=> 128,
        'ripemd128'=> 32,
        'ripemd160'=> 40,
        'ripemd256'=> 64,
        'ripemd320'=> 80,
        'whirlpool'=> 128,
        'tiger128,3'=> 32,
        'tiger160,3'=> 40,
        'tiger192,3'=> 48,
        'tiger128,4'=> 32,
        'tiger160,4'=> 40,
        'tiger192,4'=> 48,
        'snefru'=> 64,
        'snefru256'=> 64,
        'gost'=> 64,
        'gost-crypto'=> 64,
        'adler32'=> 8,
        'crc32'=> 8,
        'crc32b'=> 8,
        'crc32c'=> 8,
        'fnv132'=> 8,
        'fnv1a32'=> 8,
        'fnv164'=> 16,
        'fnv1a64'=> 16,
        'joaat'=> 8,
        'haval128,3'=> 32,
        'haval160,3'=> 40,
        'haval192,3'=> 48,
        'haval224,3'=> 56,
        'haval256,3'=> 64,
        'haval128,4'=> 32,
        'haval160,4'=> 40,
        'haval192,4'=> 48,
        'haval224,4'=> 56,
        'haval256,4'=> 64,
        'haval128,5'=> 32,
        'haval160,5'=> 40,
        'haval192,5'=> 48,
        'haval224,5'=> 56,
        'haval256,5'=> 64,
];
    private const HASH_REGEX= '/[0-9a-f]/i';

    /**
     * @SuppressWarnings(PHPMD.StaticAccess)
     */
    public static function getHashLength(String $type)
    {
        if (!array_key_exists($type, self::HASH_TYPES)) {
            throw new InvalidHashTypeException();
        }
        return self::HASH_TYPES[$type];
    }
    
    /**
     * @SuppressWarnings(PHPMD.StaticAccess)
     */
    public static function checkHashLength(String $hash, String $type)
    {
        if (self::getHashLength($type)!== strlen($hash)) {
            throw new InvalidHashException();
        }
    }

    /**
     * @SuppressWarnings(PHPMD.StaticAccess)
     */
    public static function validateHashString(String $hash)
    {
        if (!preg_match(self::HASH_REGEX, $hash)) {
            throw new InvalidHashException();
        }
    }
    /**
     * @SuppressWarnings(PHPMD.StaticAccess)
     */
    public static function validHashingAlgo(String $algo)
    {
        if (!array_key_exists($algo, self::HASH_TYPES)) {
            throw new InvalidHashTypeException();
        }
    }
}
